src/mod_zone_difficulty_scripts.cpp

```diff
@@ -536,6 +691,222 @@ class mod_zone_difficulty_worldscript : public WorldScript
+            {
+                go->AddLootMode(64);
+                sZoneDifficulty->HardmodeGameobjectsGUIDMap[instanceId].push_back(go->GetGUID());
+            }
+        }
+    }
+};
+
+class mod_zone_difficulty_dungeonmaster : public CreatureScript
+{
+public:
+    mod_zone_difficulty_dungeonmaster() : CreatureScript("mod_zone_difficulty_dungeonmaster") { }
+
+    bool OnGossipSelect(Player* player, Creature* /*creature*/, uint32 /*sender*/, uint32 action) override
+    {
+        uint32 instanceId = player->GetMap()->GetInstanceId();
+        if (action == 100)
+        {
+            LOG_ERROR("sql.sql", "Try turn on");
+            bool CanTurnOn = true;
+
+            // Forbid turning harmode on ...
+            // ...if a single encounter was completed on normal mode
+            if (sZoneDifficulty->HardmodeInstanceData.find(instanceId) != sZoneDifficulty->HardmodeInstanceData.end())
+            {
+                if (sZoneDifficulty->HardmodeInstanceData[instanceId].HardmodePossible == false)
+                {
+                    LOG_ERROR("sql.sql", "Hardmode is not Possible");
+                    CanTurnOn = false;
+                    // todo: Give Feedback
+                }
+            }
+            // ... if there is an encounter in progress
+            else if (player->GetInstanceScript() && player->GetInstanceScript()->IsEncounterInProgress())
+            {
+                LOG_ERROR("sql.sql", "IsEncounterInProgress");
+                CanTurnOn = false;
+                // todo: Give Feedback
+            }
+
+            if (CanTurnOn == true)
+            {
+                LOG_ERROR("sql.sql", "Turn on hardmode for id {}", instanceId);
+                sZoneDifficulty->HardmodeInstanceData[instanceId].HardmodeOn = true;
+                sZoneDifficulty->SaveHardmodeInstanceData(instanceId);
+                uint32 mapId = player->GetMap()->GetId();
+
+                // does the map have any creature which is supposed to have an improved lootmode?
+                if (sZoneDifficulty->HardmodeCreatureLoot.find(mapId) != sZoneDifficulty->HardmodeCreatureLoot.end())
+                {
+                    CloseGossipMenuFor(player);
+                    return false;
+                }
+
+                // find all creatures/GOs with their entry listed in sZoneDifficulty->HardmodeCreatureLoot[mapId] and AddLootMode(64)
+                std::list<Creature*> creatures;
+                for (auto entry : sZoneDifficulty->HardmodeCreatureLoot[mapId])
+                {
+                    player->GetCreaturesWithEntryInRange(creatures, 50000.0f, entry);
+                }
+                std::list<GameObject*> gameobjects;
+                for (auto entry : sZoneDifficulty->HardmodeGameobjectLoot[mapId])
+                {
+                    player->GetGameObjectListWithEntryInGrid(gameobjects, entry, 50000.0f);
+                }
```

Comment:
Just writing a reminder here.

As you are not adding any values into your lists, you are later iterating over an empty list